components/datagrid/DataGrid.js

```diff
@@ -0,0 +1,97 @@
+import React, { PureComponent } from 'react';
+import PropTypes from 'prop-types';
+import Box from '../box';
+import Cell from './Cell';
+import HeaderCell from './HeaderCell';
+import isComponentOfType from '../utils/is-component-of-type';
+import FooterRow from './FooterRow';
+import HeaderRow from './HeaderRow';
+import BodyRow from './BodyRow';
+import cx from 'classnames';
+import theme from './theme.css';
+
+class DataGrid extends PureComponent {
+  static propTypes = {
+    children: PropTypes.node,
+    className: PropTypes.string,
+    selectable: PropTypes.bool,
+  };
+
+  constructor() {
+    super(...arguments);
+
+    this.handleRowSelectionChange = ::this.handleRowSelectionChange;
+    this.handleHeaderRowSelectionChange = ::this.handleHeaderRowSelectionChange;
+
+    this.state = {
+      selectedRows: [],
+    };
+  }
+
+  handleHeaderRowSelectionChange(value) {
+    const allChildrenIndexes = React.Children.map(this.props.children, child => child.key);
+
+    if (value) {
+      this.setState({
+        selectedRows: allChildrenIndexes,
+      });
+    } else {
+      this.setState({
+        selectedRows: [],
+      });
+    }
+  }
+
+  handleRowSelectionChange(rowIndex) {
+    const { selectedRows } = this.state;
+
+    if (selectedRows.indexOf(rowIndex) > -1) {
+      this.setState({
+        selectedRows: selectedRows.filter(row => row !== rowIndex),
+      });
+    } else {
+      this.setState({
+        selectedRows: [...selectedRows, rowIndex],
+      });
+    }
+  }
```

Comment:
**Overall PP** Would like to see the functional `setState` instead of passing an object.

```js
this.setState(prevState => {
  const selectedRows = prevState.selectedRows.includes(rowIndex)
    ? prevState.selectedRows.filter(row => row !== rowIndex)
    :  [...selectedRows, rowIndex];

    return {
      ...prevState,
      selectedRows,
     };
});
```

Reply:
This is a nicer way indeed! Never thought of doing it that way.